src/dependency.rs

```diff
@@ -386,11 +386,22 @@ impl Dependency {
     }
 
     /// Modify existing entry to match this dependency
-    pub fn update_toml(&self, crate_root: &Path, item: &mut toml_edit::Item) {
-        #[allow(clippy::if_same_then_else)]
-        if str_or_1_len_table(item) {
+    pub fn update_toml(
+        &self,
+        crate_root: &Path,
+        item: &mut toml_edit::Item,
+        preserve_precision: bool,
+    ) {
+        if let Some(req) = item.as_str() {
             // Nothing to preserve
-            *item = self.to_toml(crate_root);
+            let new_item = self.to_toml(crate_root);
+            if let Some(v) = new_item.as_str() {
+                let new_req = upgrade_requirement_with_default(Some(req), v, preserve_precision)
+                    .expect("requirement to be upgraded");
+                *item = toml_edit::value(new_req);
+            } else {
+                *item = new_item;
```

Comment:
How come you are handling the format logic in `update_toml`.  imo this policy is higher level than what should exist in this area and should be handled there instead, like we do for skipping compatible.

Reply:
I agree with everything you've said. But I want to point out that that, with the current architecture, one can only push the new logic so high. The problem is that each dependency is created early: https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/bin/upgrade/main.rs#L322
But the requirement whose precision is to be preserved isn't known until much later: https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/manifest.rs#L335
In the revision I am about to push, the requirement is "cached" in the dependency. This "caching" occurs near the latter of the above. The effect is to reduce the amount of code that must be aware of the requirement. So I think it might meet your intent.

Reply:
We know the version requirement in `get_dependencies`
https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/bin/upgrade/main.rs#L271

We could put that in the `UpgradeMetadata`
https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/bin/upgrade/main.rs#L281

We could then modify the version we get back and match it to our version requirement
https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/bin/upgrade/main.rs#L434

Reply:
This approach wouldn't work with `--to-lockfile`, would it (because of where `get_upgraded` is called)?
https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/bin/upgrade/main.rs#L495
https://github.com/killercup/cargo-edit/blob/ac9cb125f466e55c38035cc8902e22b7eeff091e/src/bin/upgrade/main.rs#L518
Maybe I'm mistaken, or maybe you had an idea on how to handle that?

Reply:
I worked on a PioC of my idea but got fed up enough with the code base (like I said, this command isn't a focus right now), I did #618 .

I have rebased your changes off of that and made a quick demo of what I'm proposing here.  Its not finished.
https://github.com/epage/cargo-edit/commits/preserve
The main thing I see still lacking is preserving the formatting of the version requirement.  `semver::VersionReq` does not preserve the implicit `^`.  That should be a relatively easy refactoring
1. Create `LocalManifest::get_dependencies` based on `LocalManifest::get_dependency_versions`
2. Port `get_dependencies` in `main.rs` to use `LocalManifest::get_dependency_versions`
3. Update `old_version` to be a string instead of a `VersionReq`, re-parsing it where needed
3. Submit a PR
4. Rebase my branch off of those changes
5. Update this PR

Reply:
Thank you for going to this trouble, especially since this is not your focus right now.

If you would like, I could try to perform the steps you've laid out.

But I understand if you would prefer to handle them yourself.

Reply:
Go for it!

Reply:
Sorry to keep bother you with this, but I have a couple of questions.
1. What should `LocalManifest::get_dependencies` do that `LocalManifest::get_dependency_versions` does not?
2. The reason why `get_dependency_versions` returns an iterator is that a dependency could appear multiple places in a manifest, correct? So to accomplish 2, should (say) a vector of `table_path`s be added to `UpgradeMetadata`? I.e., the vector would contain one entry for each place the dependency appears in the manifest? (Sorry if I am completely off-base.)

Reply:
> What should LocalManifest::get_dependencies do that LocalManifest::get_dependency_versions does not?

`LocalManifest::get_dependency_versions` reports all instances for one dependency key

This new one will return all dependencies.  The code will be very similar

> The reason why get_dependency_versions returns an iterator is that a dependency could appear multiple places in a manifest, correct? So to accomplish 2, should (say) a vector of table_paths be added to UpgradeMetadata? I.e., the vector would contain one entry for each place the dependency appears in the manifest? (Sorry if I am completely off-base.)

We just care about discovering what dependencies exist at this stage.  We aren't tracking what table it is related to (its already this way).  We then find the updates for all of them and then apply all of them where they fit.

It might be worthwhile to de-duplicate dependencies though.